scispacy/umls_linking.py

```diff
@@ -0,0 +1,94 @@
+    the MedMentions dataset links to entities which have definitions 82.9% of the time. So by
+    default, we only link to entities which have definitions (typically they are more salient / cleaner),
+    but this might not suit your use case. YMMV.
+
+
+    Parameters
+    ----------
+
+    candidate_generator : `CandidateGenerator`, optional, (default = None)
+        A CandidateGenerator to generate entity candidates for mentions.
+        If no candidate generator is passed, the default pretrained one is used.
+    resolve_abbreviations : bool = True, optional (default = False)
+        Whether to resolve abbreviations identified in the Doc before performing linking.
+        This parameter has no effect if there is no `AbbreviationDetector` in the spacy
+        pipeline.
+    k : int, optional, (default = 30)
+        The number of nearest neighbours to look up from the candidate generator per mention.
+    threshold : float, optional, (default = 0.7)
+        The threshold that a mention candidate must reach to be added to the mention in the Doc
+        as a mention candidate.
+    filter_for_definitions: bool, default = True
+        Whether to filter entities that can be returned to only include those with definitions
+        in the knowledge base.
+    max_entities_per_mention : int, optional, default = 5
+        The maximum number of entities which will be returned for a given mention, regardless of
+        how many are nearest neighbours are found.
+
+    """
+    def __init__(self,
+                 candidate_generator: CandidateGenerator = None,
+                 resolve_abbreviations: bool = True,
+                 k: int = 30,
+                 threshold: float = 0.7,
+                 filter_for_definitions: bool = True,
+                 max_entities_per_mention: int = 5):
+
+        Span.set_extension("umls_ent", default=[], force=True)
+
+        self.candidate_generator = candidate_generator or CandidateGenerator()
+        self.resolve_abbreviations = resolve_abbreviations
+        self.k = k
+        self.threshold = threshold
+        self.umls = self.candidate_generator.umls
+        self.filter_for_definitions = filter_for_definitions
+        self.max_entities_per_mention = max_entities_per_mention
+
+    def __call__(self, doc: Doc) -> Doc:
+        mentions = []
+        if self.resolve_abbreviations and Doc.has_extension("abbreviations"):
+
+            for ent in doc.ents:
+                if ent._.long_form is not None:
+                    mentions.append(ent._.long_form)
+                else:
+                    mentions.append(ent)
+        else:
+            mentions = doc.ents
+
+        mention_strings = [x.text for x in mentions]
+        batch_candidates = self.candidate_generator(mention_strings, self.k)
+
+        for mention, candidates in zip(doc.ents, batch_candidates):
+            for cand in candidates[:self.max_entities_per_mention]:
```

Comment:
the max is going to have some interaction with filtering for definitions and the thresholding here. i think you probably want to keep track of all candidates and their score, and then sort them by score, and then cap to the max